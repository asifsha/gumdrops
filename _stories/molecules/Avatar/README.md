The `<Avatar>` component can display a user's photo, a brand's logo, or a flag icon for internationalization.

You can pass it an `image` src to display a photo or logo, or a `username` to display the user's first initial.

The avatar can include a dropdown menu, which is meant to be shown when the avatar is clicked on. If the avatar is expandable, pass in a `menuCallback` function that sets the state, and include the state in your component.

For the menu contents, you must pass in `menuOptions` (an array of objects of names and paths for the menu), and a `optionCallback` that will be called when the option is clicked. If you want to have a login option, you can include that in the `optionCallback`.

**Example**:

```
state {
    avatarOpen: false
}

const toggleAvatar = () => {
    this.setState({ open: !this.state.toggleAvatar });
};

const optionCallback = (path) => {
    // Example using a custom logout function if the path is logout
    // otherwise use React Router's history.push to change the path
    path === 'logout' ? logout() : browserHistory.push(path);
};

return(
    <Avatar
        open={ this.state.avatarOpen }
        username="Michele"
        optionCallback={ this.optionCallback }
        menuCallback={ this.toggleAvatar }
        menuOptions={ AvatarMenu }
    />
)
```

**Options Format**:

```
const AvatarMenu = [
    { name: 'Change Password', path: '/account/change-password' },
    { name: 'Change Brand', path: '/account/change-brand' },
    { name: 'Login to Twitter', path: '/account/twitter-login' },
    { name: 'Switch to Annotator', path: '/account/instagram-login' },
    { name: 'Logout', path: 'logout' }
];
```

**Keyboard Accessibility:**

When the avatar is in focus you can toggle the related menu opened/closed with the spacebar or enter keys.
