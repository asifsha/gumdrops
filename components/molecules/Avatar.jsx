import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import generateUID from '../utils/generateUID';
import charCodes from '../../constants/charCodes';

class Avatar extends Component {
    uid = generateUID(this);

    render() {
        const {
            open,
            menuCallback,
            optionCallback,
            username,
            img,
            menuOptions,
            className,
            style
        } = this.props;

        const rootClass = cx('gds-avatar', className, {
            'gds-avatar--menu-open': open
        });

        const regionId = `Avatar-region-${this.uid}`;
        const labelId = `Avatar-label-${this.uid}`;

        return (
            <div
                aria-controls={regionId}
                aria-expanded={open}
                aria-label={`${open ? 'Close' : 'Open'} menu options`}
                aria-pressed={open}
                className={rootClass}
                id={labelId}
                onClick={menuCallback}
                onKeyPress={event => {
                    const { type, charCode } = event;
                    if (
                        type === 'keypress' &&
                        (charCode === charCodes.SPACEBAR || charCode === charCodes.ENTER)
                    ) {
                        event.preventDefault();
                        menuCallback();
                    }
                }}
                role="button"
                style={style}
                tabIndex={0}>
                {img ? (
                    <div className="gds-avatar__image">
                        <img src={img} height="100%" alt={username} />
                    </div>
                ) : (
                    <div className="gds-avatar__image -text-center -text-tr-cap">
                        {username && username.slice(0, 1)}
                    </div>
                )}
                {menuOptions && (
                    <nav aria-labelledby={labelId} className="gds-avatar__menu" id={regionId}>
                        <ul className="gds-avatar__menu-list">
                            <li className="gds-avatar__menu-list-item -color-tx-lt-5 -ellipsis -p-h-3 -p-v-2">
                                {username}
                            </li>
                            <li className="gds-avatar__menu-list-divider" />
                            {menuOptions.map((o, i) => {
                                const _optionCallback = event => {
                                    const { type, charCode } = event;

                                    // NOTE: we can avoid keypress check if the design system
                                    // uses buttons instead of divs. Buttons currently require
                                    // additional styles to avoid breaking layout
                                    if (
                                        type === 'keypress' &&
                                        (charCode === 32 || charCode === 13)
                                    ) {
                                        event.preventDefault();
                                        event.stopPropagation();
                                        optionCallback(o.path);
                                    } else if (type === 'click') {
                                        optionCallback(o.path);
                                    }
                                };

                                return (
                                    <li
                                        key={`menu-item-${i}`}
                                        className="gds-avatar__menu-list-item -ellipsis">
                                        <div
                                            role="button"
                                            tabIndex={open ? 0 : -1}
                                            className="gds-avatar__menu-list-link"
                                            onClick={_optionCallback}
                                            onKeyPress={_optionCallback}>
                                            {o.name}
                                        </div>
                                    </li>
                                );
                            })}
                        </ul>
                    </nav>
                )}
            </div>
        );
    }
}

Avatar.displayName = 'Avatar';

Avatar.defaultProps = {
    open: false,
    menuCallback: null,
    optionCallback: null,
    username: '',
    img: '',
    menuOptions: [],
    className: '',
    style: {}
};

Avatar.propTypes = {
    open: PropTypes.bool,
    menuCallback: PropTypes.func,
    optionCallback: PropTypes.func,
    username: PropTypes.string,
    img: PropTypes.string,
    menuOptions: PropTypes.array,
    className: PropTypes.string,
    style: PropTypes.object
};

export default Avatar;
