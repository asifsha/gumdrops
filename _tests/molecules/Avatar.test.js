/* globals mount */
import React from 'react';
import Avatar from '../../components/molecules/Avatar';

const defaultProps = {
    open: false,
    menuCallback: () => {},
    optionCallback: () => {},
    username: 'Kenny',
    img: 'http://foo.com',
    menuOptions: [
        {
            name: 'test',
            path: '/test'
        },
        {
            name: 'moar',
            path: '/more'
        }
    ],
    className: 'custom-class',
    style: { width: '100px' }
};

describe('Expect <Avatar>', () => {
    it('to render', () => {
        const wrapper = mount(<Avatar {...defaultProps} />);
        expect(wrapper).toMatchSnapshot();
    });
    it('to render opened', () => {
        const props = {
            ...defaultProps,
            open: true
        };
        const wrapper = mount(<Avatar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});
