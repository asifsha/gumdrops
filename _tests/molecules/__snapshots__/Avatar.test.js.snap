// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <Avatar> to render 1`] = `
<Avatar
  className="custom-class"
  img="http://foo.com"
  menuCallback={[Function]}
  menuOptions={
    Array [
      Object {
        "name": "test",
        "path": "/test",
      },
      Object {
        "name": "moar",
        "path": "/more",
      },
    ]
  }
  open={false}
  optionCallback={[Function]}
  style={
    Object {
      "width": "100px",
    }
  }
  username="Kenny"
>
  <div
    aria-controls="Avatar-region-QXZhdGFy0"
    aria-expanded={false}
    aria-label="Open menu options"
    aria-pressed={false}
    className="gds-avatar custom-class"
    id="Avatar-label-QXZhdGFy0"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    style={
      Object {
        "width": "100px",
      }
    }
    tabIndex={0}
  >
    <div
      className="gds-avatar__image"
    >
      <img
        alt="Kenny"
        height="100%"
        src="http://foo.com"
      />
    </div>
    <nav
      aria-labelledby="Avatar-label-QXZhdGFy0"
      className="gds-avatar__menu"
      id="Avatar-region-QXZhdGFy0"
    >
      <ul
        className="gds-avatar__menu-list"
      >
        <li
          className="gds-avatar__menu-list-item -color-tx-lt-5 -ellipsis -p-h-3 -p-v-2"
        >
          Kenny
        </li>
        <li
          className="gds-avatar__menu-list-divider"
        />
        <li
          className="gds-avatar__menu-list-item -ellipsis"
          key="menu-item-0"
        >
          <div
            className="gds-avatar__menu-list-link"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={-1}
          >
            test
          </div>
        </li>
        <li
          className="gds-avatar__menu-list-item -ellipsis"
          key="menu-item-1"
        >
          <div
            className="gds-avatar__menu-list-link"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={-1}
          >
            moar
          </div>
        </li>
      </ul>
    </nav>
  </div>
</Avatar>
`;

exports[`Expect <Avatar> to render opened 1`] = `
<Avatar
  className="custom-class"
  img="http://foo.com"
  menuCallback={[Function]}
  menuOptions={
    Array [
      Object {
        "name": "test",
        "path": "/test",
      },
      Object {
        "name": "moar",
        "path": "/more",
      },
    ]
  }
  open={true}
  optionCallback={[Function]}
  style={
    Object {
      "width": "100px",
    }
  }
  username="Kenny"
>
  <div
    aria-controls="Avatar-region-QXZhdGFy1"
    aria-expanded={true}
    aria-label="Close menu options"
    aria-pressed={true}
    className="gds-avatar custom-class gds-avatar--menu-open"
    id="Avatar-label-QXZhdGFy1"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    style={
      Object {
        "width": "100px",
      }
    }
    tabIndex={0}
  >
    <div
      className="gds-avatar__image"
    >
      <img
        alt="Kenny"
        height="100%"
        src="http://foo.com"
      />
    </div>
    <nav
      aria-labelledby="Avatar-label-QXZhdGFy1"
      className="gds-avatar__menu"
      id="Avatar-region-QXZhdGFy1"
    >
      <ul
        className="gds-avatar__menu-list"
      >
        <li
          className="gds-avatar__menu-list-item -color-tx-lt-5 -ellipsis -p-h-3 -p-v-2"
        >
          Kenny
        </li>
        <li
          className="gds-avatar__menu-list-divider"
        />
        <li
          className="gds-avatar__menu-list-item -ellipsis"
          key="menu-item-0"
        >
          <div
            className="gds-avatar__menu-list-link"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
          >
            test
          </div>
        </li>
        <li
          className="gds-avatar__menu-list-item -ellipsis"
          key="menu-item-1"
        >
          <div
            className="gds-avatar__menu-list-link"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
          >
            moar
          </div>
        </li>
      </ul>
    </nav>
  </div>
</Avatar>
`;
